### PR TITLE
SNOW-765525 add compiling checker and clean up dependency

### PR DESCRIPTION
**Pull request by sfc-gh-dwang**
Initial try to add compiling checker and clean up dependency

### CODE REVIEW

File: pom.xml
Comment (sfc-gh-pbennes) [inline pom.xml:155]: Can this section be removed and the issues fixed by removing the declarations of these dependencies from pom.xml?

File: pom.xml
Comment (sfc-gh-pbennes) [inline pom.xml:283]: ?

**Review by sfc-gh-pbennes — COMMENTED** [review]
Broad comment that probably all of the `<exclusions>` blocks under all of the dependencies should be removed.

File: pom.xml
Comment (sfc-gh-dwang) [reply pom.xml:155]: if we remove this section, we will get declared but not used error, though they are used by other dependent packages

File: pom.xml
Comment (sfc-gh-dwang) [reply pom.xml:283]: we have many dependent packages use different version of this package so i have to exclude all of them and leave one version. 
Is there some other way?

**Comment by sfc-gh-dwang** [comment]
> Broad comment that probably all of the `<exclusions>` blocks under all of the dependencies should be removed.

any better way to handle different version of packages are referenced in other dependent packages?

File: pom.xml
Comment (sfc-gh-dwang) [reply pom.xml:283]: Here is the example:
`Dependency convergence error for org.slf4j:slf4j-api:1.7.21 paths to dependency are:
+-net.snowflake:snowflake-ingest-sdk:1.1.0
  +-org.slf4j:slf4j-api:1.7.21
and
+-net.snowflake:snowflake-ingest-sdk:1.1.0
  +-org.slf4j:slf4j-simple:1.7.21
    +-org.slf4j:slf4j-api:1.7.21
and
+-net.snowflake:snowflake-ingest-sdk:1.1.0
  +-org.apache.parquet:parquet-common:1.12.3
    +-org.slf4j:slf4j-api:1.7.22
and
+-net.snowflake:snowflake-ingest-sdk:1.1.0
  +-org.apache.arrow:arrow-memory-core:10.0.0
    +-org.slf4j:slf4j-api:1.7.25
and
+-net.snowflake:snowflake-ingest-sdk:1.1.0
  +-org.apache.parquet:parquet-column:1.12.3
    +-org.apache.parquet:parquet-encoding:1.12.3
      +-org.slf4j:slf4j-api:1.7.22
and
+-net.snowflake:snowflake-ingest-sdk:1.1.0
  +-org.apache.parquet:parquet-column:1.12.3
    +-org.slf4j:slf4j-api:1.7.22
`

File: pom.xml
Comment (sfc-gh-pbennes) [reply pom.xml:155]: What do you mean by other dependent packages?

https://jlbp.dev/JLBP-22 dependencies should only be specified when project source files use them.

File: pom.xml
Comment (sfc-gh-pbennes) [reply pom.xml:283]: Need to add a dependencyManagement section to set versions of transitive dependencies:
https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html

Example: https://github.com/snowflakedb/snowflake-kafka-connector/blob/0db1a86423e5d3858ba32a70420326e34caa433b/pom.xml#L79-L99

Comment (sfc-gh-dwang) [reply pom.xml:283]: I see, thanks, I will need to read the doc first too:)

File: pom.xml
Comment (sfc-gh-dwang) [reply pom.xml]: updated

File: pom.xml
Comment (sfc-gh-pbennes) [inline pom.xml:737]: Any check that's passing should be moved to the general pom.xml rather than being behind a profile so nothing regresses. Ideally that's everything and exclusions can be added for anything not currently passing.

**Review by sfc-gh-tjones — APPROVED** [review]

File: pom.xml
Comment (sfc-gh-dwang) [reply pom.xml]: updated

File: pom.xml
Comment (sfc-gh-pbennes) [inline pom.xml:536]: would be good to add comments here of the dependencies that bring in these duplicate classes

File: pom.xml
Comment (sfc-gh-pbennes) [inline pom.xml:573]: I'm assuming linkage checker is finding issues which is why the level is set WARN?

**Review by sfc-gh-pbennes — APPROVED** [review]
PR looks mergeable as-is. Room for improvement but can be done in other PRs and this PR is quite large already.

**Comment by sfc-gh-tzhang** [comment]
We plan to release 1.1.1 by end of this week which contains a few important fixes for customers, please wait for 1.1.1 to be released before merging, thanks!

**Comment by sfc-gh-tzhang** [comment]
There're a few failures you need to fix in the required merge gate, and is there any testing we cloud do to make sure this change won't break existing customers during the upgrade?

File: pom.xml
Comment (sfc-gh-dwang) [reply pom.xml]: exclude some duplicated dependency, so no more these duplicate classes

File: pom.xml
Comment (sfc-gh-dwang) [reply pom.xml:584]: linkage passed after ignore the hadoop-common and snowflake-jdbc issues

File: linkage-checker-exclusion-rules.xml
Comment (sfc-gh-dwang) [inline linkage-checker-exclusion-rules.xml:19]: @sfc-gh-pbennes Do you think if we ok to ignore the following 2 linkage issues:
conflict of hadoop-common
snowflake-jdbc shade

Comment (sfc-gh-pbennes) [reply linkage-checker-exclusion-rules.xml:19]: snowflake-jdbc issues need to be also fixed but as an intermediate step they can be ignored for now

**Comment by sfc-gh-dwang** [comment]
@sfc-gh-tzhang @sfc-gh-tjones can you help check the snyx failure? I have no access to that app.

**Comment by sfc-gh-tzhang** [comment]
> @sfc-gh-tzhang @sfc-gh-tjones can you help check the snyx failure? I have no access to that app.

I don't either, @sfc-gh-jfan any idea?

**Comment by sfc-gh-tjones** [comment]
> @sfc-gh-tzhang @sfc-gh-tjones can you help check the snyx failure? I have no access to that app.

Sorry I don't have access - looks like Toby already pinged the appropriate owner.

**Review by sfc-gh-tjones — APPROVED** [review]

**Comment by sfc-gh-dwang** [comment]
Snyx failure due to hadoop-common license check.
Although hadoop-common is apache 2.0 license, it has dependency on com.sun.jersey which is double licensed on GPL2 and apache 2.0. Checked with Max Fei that double license should be fine so we should be fine with the license issue.

@sfc-gh-jfan can you help to pass the license check as we discussed in the slack?

File: pom.xml
Comment (sfc-gh-tzhang) [inline pom.xml:553]: remove?

**Review by sfc-gh-tzhang — APPROVED** [review]
LGTM as long as there is no behavior change for customers, thanks!

File: linkage-checker-exclusion-rules.xml
Comment (sfc-gh-dwang) [reply linkage-checker-exclusion-rules.xml]: removed

File: pom.xml
Comment (sfc-gh-dwang) [reply pom.xml]: removed

**Comment by sfc-gh-pbennes** [comment]
> @sfc-gh-dwang This PR removed `slf4j-simple` from the test scope, so we don't have test logs now. Was it on purpose or an overlook?

overlook

File: linkage-checker-exclusion-rules.xml
Comment (sfc-gh-pbennes) [reply linkage-checker-exclusion-rules.xml:19]: Yes, problems with shading and relocation @sfc-gh-lsembera AFAICT. I think someone's going to work on addressing these in follow-up PRs.

**Comment by sfc-gh-dwang** [comment]
@sfc-gh-lsembera likely it was removed unintentionally, can you help add back or I shall create a new PR?

**Comment by sfc-gh-lsembera** [comment]
@sfc-gh-dwang I will add it back in https://github.com/snowflakedb/snowflake-ingest-java/pull/397

**Pull request merged**
